[Frontend] Implement missing book GUI (functionality, assets, layout adjustments)

kida31 opened the pull request:
# Book GUI is now hooked up!
## Structure of book GUI

- BookView
    - PlayerPage
        - Left (Inventories)
            - TabButtons
            - InventoryView
        - Right (PlayerInfo)
    - Visuals
    - WikiPage
        - Left (List of items)
            - three Category buttons (jumps to first item of that type)
            - list of items in item database
        - Right (Article about item)

## Features
- Switch through PlayerPage and Wiki via Left and Right Triggers (not the front shoulder buttons, the back ones).
- PlayerPage
    - Switch through item categories via Left and Right Bumpers (front shoulder buttons).
    - Shows inventory on the left.
        - In the inventory, pick up items via (A). And deposit items that way too. Picking up/Dropping single items will be added soon.
    - On the right hand side you can see current Faith progress. The petals fill up as you collect more faith.
        - The achievement-style information on the bottom half is placeholder and will be implemented soon.
- WikiPage
    - Right hand side has details IItemStack.WikiDescription and stats (not implemented)
    - Right hand side bottom has list of related items. Click on them to jump to them in the list
- Focus
    - PlayerPage jumps to first item in inventory per default
    - WikiPage jumps to first item in list per default

# Changelog (in total)
- EventBus
    - Pulled most GUI events from event bus into GUI itself
    - Add Inventory change event
    - Add Inventory open event
- Game.cs
    - Adds a service locator like object. This is an experimental concept to centralize commonly accessed objects like player or inventory.
    - Single relevant classes may register themselves via ProvideX(this)
    - Other objects may retrieve those via GetX()
    - May be replaced be "GetTree().GetFirstNodeInGroup"
- Input
    - Directional Input (WASD and DPad) now properly map onto UI navigation
    - Button South (A/X on Gamepad and G on keyboard) now properly map to UI "accept" (space/enter)
- GUI
    - The Book is now hooked up with features! (Renamed some older components and moved them)
    - "Selector" (UIFocus) icon that shows currently selected Control Node
    - GlobalTooltip
    - Any Controls implementing `ITooltipable` will show a tooltip when focused for a while
    (does currently not get triggered on-hover like you would expect for mouse)
    - "Hooks" (Add these as child node to some parent object, i do not know how to call these)
        - AutoExpandHook: Node that makes parent auto expand while keeping aspect ratio (Looking for better name, thanks)
        - CopySizeHook: Node that makes parent mimic size of some Target node
        - FollowControlHook: Parent position follows some target control node
        - FollowCursorHook: Makes parent follow cursor position
        - FollowFocusedHook: Makes parent follow focused control element
- Inventory/Items
    - Removed ItemStack(a, b, c, d, e, f,....) constructor which had a million arguments, because changes here would trigger too many follow up changes.
    - Add some contentChanged events to collections
    - Remove old CategoryInventory
    - Add global "CursorInventory" (item in cursor-hand)

- Player
    - Added inventory to player, and pickup listener
- RandomStock
    - Add method to get random items from item database instead of just placeholders
- File renamed
    - This triggers a dependency changes in a few scene files

# Bugfixes
- Input, fix shoulder button mapping for book
- Fix vending machine UI reacting to "close" input although it is not even open, causing the Book to freeze up
- Remove broken item id references to non-existant stuff

kida31 commented:
- [x] Item label inventory page not auto expanding
- [x] H to close book does not work
- [x] Mouse follow for itemcursor only works outside of UI
- [x] Book only works on UI Confirm instead of A
- Layout breaks when aspect ratio is not locked -> lower priority since we lock aspect ratio
- weird font thickness after certain characters (hyphen?) -> visual font issue, lower priority, will be handle separately
- [x] Item Icons in wiki list do not scale

felix-schindler commented on Common/Logger.cs:
```diff
@@ -75,6 +70,11 @@ private void Log(LogLevel level, string message)
 				LogLevel.Error => BBColor.Red.Apply(logMessage),
 				_ => logMessage,
 			};
+			if (LogLevel.Error == level)
```
Ist es Absicht, dass Warnings hier nicht mehr extra gehandelt werden?

kida31 replied:
Das soll den "Debugger" tab etwas entlasten, damit, falls da mal was steht, es betont wird, dass man sich darum kuemmern soll.

PushWarning und PushError werfen die Meldung naemlich in diesen Debugger, sind aber relativ wenig hilfreich wenn man die ueber Logger aufruft.
PushWarning und PushError zeigen aehnlich wie bei einer Exception den ganzen CallStack. (gut)
Wenn man darauf klickt wird man jedoch nur zum "PushX" call im Logger weitergeleitet. (useless)

Beim Beobachten von Troubleshooten sah es aus als wuerde es mehr verwirren als helfen bei kleinen Dingen (Warnings). Bei Errors habe ich es mal behalten, da diese seltener sein sollten und da mehr Info mMn gut ist.

felix-schindler commented on project.godot:
```diff
@@ -42,10 +42,12 @@ CurrencyFaithOfficer="*res://Features/Systems/CurrencyFaithOfficer.cs"
 
-window/size/viewport_width=1280
-window/size/viewport_height=720
+window/size/viewport_width=1080
```
Sind das die finalen Viewport-Einstellungen auf die sich geeinigt wurde?

kida31 replied:
Nicht die finalen. Das (jetzige) ist die Referenz-Aufloesung in der Aesprite gearbeitet hat. (1080p x 200%).

Sollte der Change erstmal aufgeschoben werden?

kida31 replied:
Die finale Aufloesung wird wahrscheinlich 360p sein statt 540p. Ist aber noch pending

felix-schindler replied:
Okay, denke das passt erstmal, gerade wenn es nochmal überarbeitet wird.

felix-schindler commented on Features/GUI/Book/Inventories/InventoryItemView.cs:
```diff
@@ -0,0 +1,52 @@
+		else
+		{
+			// Placeholders: 
+			_displayItemName.Text = itemStack.Name ?? "Placeholder_ItemName";
```
~~Soweit ich sehe ist `ItemStack.Name` vom Typ String und wird sogar mit einem leeren String initialisiert.
In JavaScript ist `??` ein Null**ish** coalescing operator. Ein leerer String wäre Nullish.
Funktioniert das genauso auch in C# oder ist es dort ein richter NULL coalescing operator, der nicht bei einem leeren String anspringt?~~

Edit: Ich glaube du meintest eigentlich `itemStack?.Name` aber weil davor schon ein check `!= null` ist, kann das `?? "Placeholder_ItemName";` hier wegfallen. Ich denke `itemStack.Name` ist hier immer ein String.

kida31 replied:
Changed

felix-schindler commented on Features/GUI/Book/Inventories/InventoryTabsView.cs:
```diff
@@ -0,0 +1,202 @@
+
+	private readonly Dictionary<InventoryCategoryTab, Action> _buttonActions = new();
+
+	private int _currentTab = -1;
```
Soll das hier wirklich mit -1 initialisiert werden, wenn im setter eigentlich das Minimum 0 (`Math.Clamp(value, 0, _inventories.Count);`) angegeben ist?

kida31 replied:
Es sollte als Placeholder fuer "not set yet" stehen, weil urspruenglich beim initial setup sich etwas aufgehangen hat.

Logik ist nochmal ueberarbeitet und sollte kein Problem mehr sein

felix-schindler commented on Features/GUI/Book/Inventories/InventoryView.cs:
```diff
@@ -0,0 +1,139 @@
+		CursorInventory.Instance.ReturnPickUp();
+	}
+
+	private void OnVisibilityOn()
```
Sollte diese leere Funktion besser entfernt werden?

kida31 replied:
Removed

felix-schindler commented on Features/GUI/Book/PlayerStuff/PlayerDetailsView.cs:
```diff
@@ -0,0 +1,31 @@
+
+	public void SetPlayerDetails(object _)
+	{
+		label1.Text = "Total Gold: 0";
+		label2.Text = "Items Sold: 0";
+		label3.Text = "Plants harvested: 0";
+		label4.Text = "Plants died: 0";
```
Sollten diese String vielleicht. lieber. Aussehen wie `"Total Gold: {:VALUE:}"` und dann ein Setter haben in dem man (pseudo code) `UpdateTotalGold(int i) { str.Replace("{:VALUE:}", i) }`.

kida31 replied:
Platzhalter ersetzt mit korrekten Setter calls in ready.
Ein mega-setter ersetzt durch viele Mini Setter die ausdrucksstaerker sind.

felix-schindler commented on Features/Player/Player.cs:
```diff
@@ -31,7 +28,7 @@ public partial class Player : CharacterBody2D
 	private AnimatedSprite2D _animatedSprite2D;
 	private PlayerStateMachine _stateMachine;
-	private BigInventory _inventory = new(16);
+	private BigInventory _inventory = new(20);
```
Ist die Value in `BigInventory: this(20)` nicht sowieso schon gesetzt? Wenn man sie hier nochmal setzt ist es besser oder soll der Spieler immer das default BigInventory haben?

kida31 replied:
Removed empty constructor. Und redundant second assignment entfernt

felix-schindler commented on Features/Tools/SeedBag.cs:
```diff
@@ -21,7 +21,7 @@ public SeedBag(float radius, float range, float channelingTime) : base(radius, r
 	{
 		var inventory = user.Inventory.GetInventory(ItemCategory.Seed);
-		CurrentSeedItem = inventory.FirstOrDefault();
+		CurrentSeedItem = inventory.FirstOrDefault() as ItemStack; // This might break if ItemStack is not properly updated (independent of IItemstack)
```
Ist das ein `TODO: ` Kommentar oder lassen wir das so? :D

kida31 replied:
Wir lassen das so bis es breakt. oder zum naechsten Seed related change 😄

felix-schindler reviewed (commented):
Alles in allem ist der Code sehr gut. Einige kleinliche Kommentare habe ich da gelassen, aber an sich ist dort nichts dabei was einen Merge wirklich abhalten sollte.

kida31 commented on Common/Attributes/MayBeBuggyAttribute.cs:
Renamed to `Unstable`

kida31 commented:
Description updated

Sui-nyan reviewed (approved):
🆗

kida31 commented:
Resolving conversations since PR is approved.

kida31 commented:
Did another backmerge (and redid some changes), still works.
Mostly commits to re-do stuff.
Changed viewport to 360p @Sui-nyan 
Cleaned up some debug prints.

kida31 commented:
> Did another backmerge (and redid some changes), still works. Mostly commits to re-do stuff. Changed viewport to 360p @Sui-nyan Cleaned up some debug prints.
> 
> #trustmebro

i did indeed *break*. #176
